src/Data/DefaultDataProvider.php

```diff
@@ -176,40 +177,67 @@ public function enableDefaultUuidGenerator()
      *
      * @throws DcGeneralRuntimeException When no source has been defined.
      */
-    public function setBaseConfig(array $arrConfig)
+    public function setBaseConfig(array $config)
     {
         // Check configuration.
-        if (!isset($arrConfig['source'])) {
+        if (!isset($config['source'])) {
             throw new DcGeneralRuntimeException('Missing table name.');
         }
 
-        if (isset($arrConfig['database'])) {
-            if (!($arrConfig['database'] instanceof Database)) {
-                throw new DcGeneralRuntimeException('Invalid database.');
+        // CHANGE TO CONNECTION START
+        // This is the fallback for deprecated contao database.
+        if (isset($config['database'])) {
+            @trigger_error(
+                'Config key database is deprecated use instead connection. Fallback will be dropped.',
+                E_USER_DEPRECATED
+            );
+
+            if (!isset($config['connection'])) {
+                $config['connection'] = $config['database'];
+            }
+
+            unset($config['database']);
+        }
+
+        if (isset($config['connection'])) {
+            if ($config['connection'] instanceof Database) {
+                @trigger_error(
+                    'Contao database is deprecated use doctrine dbal connection. Fallback will be dropped.',
+                    E_USER_DEPRECATED
+                );
+
+                $config['connection'] = System::getContainer()->get('database_connection');
```

Comment:
I think that's wrong to use the default connection here. If someone passes a custom connection it's probably not the default one. Maybe you could fetch the connection using reflection from the database object. 

You may have a look at MM 2.1, @discordier does it there somewhere...

Reply:
Something along the lines of:
```php
        // BC layer - we used to accept a Contao database instance here.
        if ($dataBase instanceof Database) {
            @trigger_error(
                '"' . __METHOD__ . '" now accepts doctrine instances - ' .
                'passing Contao database instances is deprecated.',
                E_USER_DEPRECATED
            );
            $reflection = new \ReflectionProperty(Database::class, 'resConnection');
            $reflection->setAccessible(true);

            $dataBase = $reflection->getValue($dataBase);
        }
        if (null === $dataBase) {
            @trigger_error(
                'You should pass a doctrine database connection to "' . __METHOD__ . '".',
                E_USER_DEPRECATED
            );
            $dataBase = System::getContainer()->get('database_connection');
        }
```

Reply:
Has been change